code/modules/projectiles/guns/flamer.dm

```diff
@@ -544,19 +544,33 @@
 	var/obj/screen/ammo/A = user.hud_used.ammo
 	A.update_hud(user)
 
-/obj/item/weapon/gun/flamer/marinestandard/Fire(atom/target, mob/living/user, params, reflex)
+/obj/item/weapon/gun/flamer/marinestandard/Fire()
 	if(active_attachable && istype(active_attachable, /obj/item/attachable/hydro_cannon) && (world.time > last_use + 10))
-		extinguish(target,user) //Fire it.
+		INVOKE_ASYNC(src, .proc/extinguish, target, gun_user) //Fire it.
 		water_count -=7//reagents is not updated in this proc, we need water_count for a updated HUD
 		last_fired = world.time
 		last_use = world.time
-		var/obj/screen/ammo/A = user.hud_used.ammo
-		A.update_hud(user)
-		return
-	if(user.skills.getRating("firearms") < 0 && !do_after(user, 1 SECONDS, TRUE, src))
+		var/obj/screen/ammo/A = gun_user.hud_used.ammo
+		A.update_hud(gun_user)
 		return
+	if(gun_user.skills.getRating("firearms") < 0)
+		switch(windup_checked)
+			if(WEAPON_WINDUP_NOT_CHECKED)
+				INVOKE_ASYNC(src, .proc/do_windup)
```

Comment:
We can't sleep in a SIGNAL_HANDLER, so we invoke a sleeping proc that will call Fire again once the windup is done